misc/python/materialize/mzexplore/clone.py

```diff
@@ -0,0 +1,155 @@
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+"""
+Utilities to extract data from a Materialize catalog for exploration purposes.
+"""
+
+import string
+import textwrap
+from contextlib import closing
+from typing import TextIO
+
+from pg8000.dbapi import DatabaseError
+
+from materialize.mzexplore import sql
+from materialize.mzexplore.common import ItemType, info, warn
+
+
+def defs(
+    ddl_out: TextIO,
+    cmp_out: TextIO,
+    database: str,
+    schema: str,
+    cluster: str,
+    object_ids: list[str],
+    db_port: int,
+    db_host: str,
+    db_user: str,
+    db_pass: str | None,
+    db_require_ssl: bool,
+    mzfmt: bool,
+) -> None:
+    """
+    Generate a DDL script that clones object_ids and dependencies that live in
+    the same cluster into a different cluster.
+
+    The catalog items included in the script will be seeded from the .
+    """
+
+    database_new = sql.identifier(database, True)
+    schema_new = sql.identifier(schema, True)
+    cluster_new = sql.identifier(cluster, True)
+
+    with closing(
+        sql.Database(
+            port=db_port,
+            host=db_host,
+            user=db_user,
+            database=None,
+            password=db_pass,
+            require_ssl=db_require_ssl,
+        )
+    ) as db:
+        output_template = string.Template(
+            textwrap.dedent(
+                """
+                -- original id: $id
+                $create_sql
+                """
+            ).lstrip()
+        )
+
+        # Extract dependencies
+        # --------------------
+
+        if len(object_ids) == 0:
+            msg = "At least one object_id necessary"
+            raise ValueError(msg)
+
+        old_clusters = list(db.object_clusters(object_ids))
+        if len(old_clusters) != 1:
+            msg = f"Cannot find unique old cluster for object ids: {object_ids}"
+            raise ValueError(msg)
+
+        [cluster_old] = old_clusters
+
+        if cluster_old["name"] == cluster:
+            msg = f"Old and new clusters can't have the same name: {cluster}"
+            raise ValueError(msg)
+
+        items_sub = dict()
+        for item in db.clone_dependencies(object_ids, cluster_old["id"]):
+            item_database = sql.identifier(item["database"], True)
+            item_schema = sql.identifier(item["schema"], True)
+            item_name = sql.identifier(item["name"], True)
+            fqname = f"{item_database}.{item_schema}.{item_name}"
+
+            try:
+                item_type = ItemType(item["type"])
+            except ValueError:
+                warn(f"Unsupported item type `{item['type']}` for {fqname}")
+                continue
+
+            show_create_query = item_type.show_create(fqname)
+            if show_create_query is None:
+                continue
```

Comment:
Maybe we want to assert or return an error here instead? Are there legitimate reasons why `show_create_query` could be `None`?

Reply:
Yes, makes sense.